data/padics/hensel.lean

```diff
@@ -0,0 +1,503 @@
+/-
+Copyright (c) 2018 Robert Y. Lewis. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Robert Y. Lewis
+
+A proof of Hensel's lemma on ℤ_p, roughly following Keith Conrad's writeup:
+http://www.math.uconn.edu/~kconrad/blurbs/gradnumthy/hensel.pdf
+-/
+
+import data.padics.padic_integers data.polynomial data.nat.binomial data.real.cau_seq_filter analysis.limits tactic.ring
+
+noncomputable theory
+
+local attribute [instance] classical.prop_decidable
+
+section identities
+open polynomial
+
+lemma pow_add_expansion {α : Type*} [comm_semiring α] (x y : α) : ∀ (n : ℕ),
+  {k // (x + y)^n = x^n + n*x^(n-1)*y + k * y^2}
+| 0 := ⟨0, by simp⟩
+| 1 := ⟨0, by simp⟩
+| (k+2) :=
+  begin
+    cases pow_add_expansion (k+1) with z hz,
+    rw [_root_.pow_succ, hz],
+    existsi (x*z + (k+1)*x^k+z*y),
+    simp [_root_.pow_succ], ring -- expensive!
+  end
+
+variables {α : Type*} [decidable_eq α] [comm_ring α]
+
+private lemma poly_binom_aux1 (x y : α) (e : ℕ) (a : α) :
+  {k : α // a * (x + y)^e = a * (x^e + e*x^(e-1)*y + k*y^2)} :=
+begin
+  existsi (pow_add_expansion x y e).val,
+  congr,
+  apply (pow_add_expansion _ _ _).property
+end
+
+private lemma poly_binom_aux2 (p : polynomial α) (x y : α) :
```

Comment:
This is definitely a matter of taste, but in a file related to the p-adics I would prefer if polynomials were not called `p` but rather something like `f`.

Reply:
Fair point! This was moved from another file. I'm still not sure if it belongs here or somewhere else.